bentoml/yatai/db/stores/lock.py

```diff
@@ -0,0 +1,54 @@
+import enum
+
+from sqlalchemy import UniqueConstraint, Column, Integer, Enum
+from bentoml.exceptions import LockUnavailable
+from bentoml.yatai.db import Base
+
+
+class LOCK_STATUS(enum.Enum):
+    read_lock = 1
+    write_lock = 2
+
+
+class Lock(Base):
+    __tablename__ = 'locks'
+    __table_args__ = tuple(UniqueConstraint('resource_id', name='_resource_id_uc',))
+    id = Column(Integer, primary_key=True)
+    resource_id = Column(Integer, nullable=False)
+    lock_status = Column(Enum(LOCK_STATUS))
+
+
+class LockStore(object):
+    @staticmethod
+    def _find_lock(sess, resource_id):
+        lock_obj = sess.query(Lock).filter_by(resource_id=resource_id).first()
+        return lock_obj
+
+    @staticmethod
+    def acquire(sess, lock_type, resource_id):
+        lock = LockStore._find_lock(sess, resource_id)
+
+        # no lock found; free to acquire
+        if not lock:
+            lock = Lock()
+            lock.lock_status = lock_type
+            lock.resource_id = resource_id
+            sess.add(lock)
+            return True
+
+        # acquire read lock
+        if lock_type == LOCK_STATUS.read_lock:
+            if lock.lock_status == LOCK_STATUS.write_lock:
+                raise LockUnavailable("Failed to acquire read lock, write lock held")
+
+            # read lock acquisition success, no change required
+            # as current state is already read_lock
+            return True
+        else:
+            # can't acquire read/write lock when write lock is already held
+            raise LockUnavailable("Failed to acquire write lock, lock held")
+
+    @staticmethod
+    def release(sess, resource_id):
+        lock = LockStore._find_lock(sess, resource_id)
```

Comment:
It doesn't verify the owner of the lock. A worker may release the lock held by other processes.

Reply:
I was thinking that we don't expose this acquire/release API to users and only lock using the context manager in `yatai/locking/lock.py`. That way, we don't need to store lock owner in the table.